File: 0279-perfect-squares/0279-perfect-squares.cpp
const int MAX = 1e4 + 5;
int memory[MAX];
class Solution
{
public:
    int perfectSquares(int n)
    {
        if (n <= 0)
            return 0;

        auto &ret = memory[n];
        if (ret != -1)
            return ret;

        ret = n;
        for (int i = 1; i < n; ++i)
        {
            if (i * i > n)
                break;
            ret = min(ret, 1 + perfectSquares(n - i * i));
        }
        return ret;
    }
    int numSquares(int n)
    {
        memset(memory, -1, sizeof(memory));
        return perfectSquares(n);
    }
};

File: 0279-perfect-squares/0279-perfect-squares.java
import java.util.Arrays;

class Solution {
    public int numSquares(int n) {
        int[] dp = new int[n + 1];
        for (int num = 1; num < n + 1; num++) {
            dp[num]=n; // worst case ones 
            for (int numSquare = 1; numSquare <= num; numSquare++) {
                int square = numSquare * numSquare;
                if (num - square < 0) break;
                dp[num] = Math.min(dp[num], 1 + dp[num - square]);
            }
        }
        return dp[n];
    }
}